### ISSUE TITLE
Platform OS Errno 5 Input/Output Error

### ISSUE THREAD

**Issue by BiagioFesta** [quinn-rs/quinn#1190]
Sporadically, during a QUIC session the library prints the following error message:

```
WARN quinn::platform: sendmsg error: Os { code: 5, kind: Uncategorized, message: "Input/output error" }, Transmit: { destination: [::ffff:xx.xx.xx.xx]:4172, src_ip: None, enc: None, len: 1481, segment_size: Some(1200) }
```

If I am not wrong, this is a failure in libc `sendmsg` (EIO-5).

I don't find useful information around about this specific error which seems to be quite general.

Does somebody have some hints and can point me in the right direction so I can start digging a little bit in it?

**Comment by Ralith** [comment]
The ultimate source of truth on that is your kernel. Unfortunately, the Linux UDP stack seems to have a number of error conditions with poor/no documentation. A good place to start would be trying to reproduce the error under controlled circumstances, i.e. call `sendmmsg` on a socket configured similarly to quinn's (see `fn init` in `quinn/src/platform/unix.rs`) and similar cmsgs (see `fn prepare_msg` in the same file). Note that observed behavior may depend on kernel UDP buffer and routing table state. A strace capture of the error occurring is an alternative to reading the code to try to reproduce the exact syscall. Someone more experienced with kernel debugging might be able to advise further.

Let us know if you learn anything!

**Comment by Ralith** [comment]
Also, I'm assuming that your connection is otherwise working fine. If not, there are a few common pitfalls like address family mismatch for single-stack sockets.

**Comment by BiagioFesta** [comment]
Ok I think I've found something.

Dislaimer: In my development environment client and server are on the same host.


The send I/O error seems to be caused by GSO on UDP.
Bisecting the code my environment fails using this commit: https://github.com/quinn-rs/quinn/commit/f1c7c561b9a2474c0a11634263474fb2ef7e9c6c

(The code right before works fine).

On that particular commit the quinn endpoint is dropped and the server stops working.

Now, on main branch, probably the issue is mitigated by: https://github.com/quinn-rs/quinn/commit/4e95796240c3f3fd19c9eec44aaf529767c3128d where is error has been "lowered" to a warning.

Maybe there are some other correlated development between those two commits I did not investigated yet.

**Comment by Ralith** [comment]
Interesting; can you confirm that all GSO transmits fail? What's your kernel version?

**Comment by BiagioFesta** [comment]
I can confirm the error is due to the GSO, but not all sends with GSO fail.

With a reference to the code: [send](https://github.com/quinn-rs/quinn/blob/main/quinn/src/platform/unix.rs#L192),

`libc::sendmmsg` *sometimes* fails with `errno=5` (i.e. Input/Ouput error), that only happens when *at least one* `Transmit` contains more fragments (GSO), that is, [pub segment_size: Option<usize>](https://github.com/quinn-rs/quinn/blob/main/quinn-proto/src/lib.rs#L299) is `some`.

In some cases, all successive calls to `sendmmsg` will fail resolving into a critical application failure.

Disabling GSO in the code (forcing [`let max_datagrams = 1;`](https://github.com/quinn-rs/quinn/blob/main/quinn/src/connection.rs#L809)) of course "fix" the issue.

I am testing running client and server on my localhost; kernel: `5.4.0-84-generic`; using a WiFi NIC.

**Comment by Ralith** [comment]
To clarify, do calls fail if and only if `segment_size` is `Some`, or do some cases where `segment_size` is `Some` still succeed?

Perhaps it fails iff there's a mixture of segmented and non-segmented transmits in the same call?

**Comment by BiagioFesta** [comment]
It's not a necessary and sufficient condition.
In other words, there are some cases indeed where `segment_size` is `Some` (for some transmit of the "batch") and the send succeeds.

I don't think it is the mix because I can reproduce the error even when `transmits: &[Transmit]` has length `1`.

**Comment by Matthias247** [comment]
> Now, on main branch, probably the issue is mitigated by: [4e95796](https://github.com/quinn-rs/quinn/commit/4e95796240c3f3fd19c9eec44aaf529767c3128d) where is error has been "lowered" to a warning.

Is it mitigated - which means the retransmit succeeds? Or will it always fail with the same payload.

However since retransmits are built on application level they actually contain different payload (at least reencrypted, but they might also contain totally different frames). 

It would be interesting to know whether some particular packets constellations with GSO are causing the issue, or whether its intermittent failures independent of the payload - e.g. if the network stack has no hardware GSO support, and the software emulation would only allow to enqueue a part of the payload. 

You could check whether its payload related by moving the `send` implementation to not drop the `Transmit` on `EIO` but keep it around and retry.

**Comment by Matthias247** [comment]
Are you binding your endpoint to a IPv4 or IPv6 socket?

If its the former, can you try the latter?

**Comment by BiagioFesta** [comment]
Sorry, I have been having a busy week. 

I am going to find time to investigate deeper by the end of the week.

For now, what I can say is:

> Is it mitigated - which means the retransmit succeeds? Or will it always fail with the same payload.

From my experiments, it was both. Sometimes the call went into a endless loop of failure of this error and the connection went in timeout because no more traffic.

However, thank you for the suggestions! I think it would be interesting indeed to check if there is some kind of correlation with the nature of the packet.

**Comment by BiagioFesta** [comment]
By the way, do you think there is a workaround in order to bypass GSO in case this happens on a production environment?

I tried `ethtool -K ${INTF} gso off` but quinn [here](https://github.com/quinn-rs/quinn/blob/main/quinn/src/platform/unix.rs#L527) still detects GSO enabled, maybe I need to restart (hard reboot) the machine on the change?

**Comment by Ralith** [comment]
Should be fixed by #1210. Feel free to reopen if not.
